cql3/describe_processor.cc

```diff
@@ -0,0 +1,241 @@
+public:
+    listing(std::function<std::vector<schema_element>(const std::vector<keyspace_element>&)> provider) 
+        : _element_provider(std::move(provider)) {}
+
+    virtual std::vector<lw_shared_ptr<column_specification>> metadata() override {
+        return list_metadata();
+    }
+
+protected:
+    virtual std::vector<schema_element> describe(const service::client_state& client_state, const std::vector<keyspace_element>& keyspaces) override {
+        auto ks = client_state.get_raw_keyspace();
+        return _element_provider(ks.empty() ? keyspaces : std::vector<keyspace_element>{get_keyspace(ks, keyspaces)});
+    }
+
+    virtual std::vector<bytes_opt> element_to_row(const schema_element& element) override {
+        return {
+            to_bytes(get_element_keyspace(element)),
+            to_bytes(get_element_type(element)),
+            to_bytes(get_element_name(element))
+        };
+    }
+};
+
+class element : public describe_processor {
+private:
+    seastar::sstring _keyspace;
+    seastar::sstring _name;
+    std::function<std::vector<schema_element>(const keyspace_element&, const seastar::sstring&)> _element_provider;
+
+public:
+    element(seastar::sstring keyspace, seastar::sstring name, 
+        std::function<std::vector<schema_element>(const keyspace_element&, const seastar::sstring&)> provider)
+        : _keyspace(keyspace), _name(name), _element_provider(std::move(provider)) {}
+
+    virtual std::vector<lw_shared_ptr<column_specification>> metadata() override {
+        return element_metadata();
+    }
+
+protected:
+    virtual std::vector<schema_element> describe(const service::client_state& client_state, const std::vector<keyspace_element>& keyspaces) override {
+        auto ks = _keyspace.empty() ? client_state.get_raw_keyspace() : _keyspace;
+        if (ks.empty()) {
+            throw exceptions::invalid_request_exception("No keyspace specified and no current keyspace");
+        }
+        
+        return _element_provider(get_keyspace(ks, keyspaces), _name);
+    }
+
+    virtual std::vector<bytes_opt> element_to_row(const schema_element& element) override {
+        return {
+            to_bytes(get_element_keyspace(element)),
+            to_bytes(get_element_type(element)),
+            to_bytes(get_element_name(element)),
+            to_bytes(get_element_string(element, this->_internals, false))
+        };
+    }
+};
+
+void describe_processor::set_internals(bool internals) {
+    _internals = internals;
+}
+
+std::unique_ptr<result_set> describe_processor::execute(query_processor& qp, service::query_state& state, const query_options& options) {
+    auto keyspaces = boost::copy_range<std::vector<keyspace_element>>(
+        qp.db().get_keyspaces() | boost::adaptors::transformed([] (auto& ks) { return ks.metadata(); })
+    );
+    boost::range::sort(keyspaces, element_comparator);
+    auto elements = describe(state.get_client_state(), keyspaces);
+
+    //TODO: Pagination
```

Comment:
Please remember what the Cassandra documentation says should happen if the schema changed during pagination:

> If a schema changes is detected between  two pages the query will fail with an ``InvalidRequestException``. It is safe to retry the whole ``DESCRIBE``   statement after such an error.

You can fairly easily write a cql-pytest test for this case - run a DESCRIBE query with a tiny page size, get the first page, modify the schema and then try to continue the paging.

Reply:
Now the implementation is using internal query to virtual table, so it should do the pagination. But I'll write a cql-pytest for this